File: splitNum.java
public class splitNum {
    public static int[] split(int n) {
        StringBuilder left = new StringBuilder();
        StringBuilder right = new StringBuilder();
        boolean flag = true;

        for (char bit : Integer.toBinaryString(n).toCharArray()) {
            if (bit == '0') {
                left.append(bit);
                right.append(bit);
            } else if (bit == '1') {
                if (flag) {
                    left.append(bit);
                    right.append('0');
                } else {
                    right.append(bit);
                    left.append('0');
                }
                flag = !flag;
            }
        }

        return new int[] { Integer.parseInt(left.toString(), 2), Integer.parseInt(right.toString(), 2) };
    }
    public static int[] split2 (int n) {
        int numBits = Integer.SIZE - Integer.numberOfLeadingZeros(n);
        
        int evenMask = 0;
        for (int i = 0; i < numBits; i += 2) {
            evenMask |= (1 << i);
        }

        int oddMask = evenMask << 1;

        int a = n & evenMask;
        int b = n & oddMask;
        
        return new int[] { a, b };
    }

    public static void main(String[] args) {
        int n = 13;  
        int[] result = split(n);
        System.out.println(Integer.toBinaryString(result[0]) + " + " + Integer.toBinaryString(result[1]));
    }
}

File: splitNum.py
def split(n: int):
    left, right = '', ''
    flag = True
    for i in bin(n)[2:]:
        if i == '0':
            left += i
            right += i
        elif i == '1':
            if flag == True:
                left += i
                right += '0'
            else:
                right += i
                left += '0'
            flag = ~flag

    return left, right

print(split(6))
print(split(7))
print(split(13))
print(split(int('110110101', 2)))

def split_bitwise(n):
    num_bits = n.bit_length()
    
    even_mask = 0
    for i in range(0, num_bits, 2):
        even_mask |= (1 << i)
    odd_mask = even_mask << 1
    
    a = n & even_mask
    b = n & odd_mask
    
    return a, b
print(split_bitwise(6))
print(split_bitwise(7))
print(split_bitwise(13))
